### PR TITLE
✨ v3 (feature): request binding

### PR DESCRIPTION
**Pull request by trim21**
**Please provide enough information so that others can review your pull request:**

related to https://github.com/gofiber/fiber/issues/1821 and https://github.com/gofiber/fiber/issues/2002

**Explain the *details* for making this change. What existing problem does the pull request solve?**

<del>this PR add a function decorator style bind method to provide high speed and low alloc request binding.</del>

this PR add a high performance and low alloc binding implement.

some changes: 

1. `data[]=john&data[]=doe` won't bind to `query:"data"`, only bind to `query:"data[]"` and can't bind query like `data[0]=john&data[1]=doe`
2. you can't used multiple tag together, only one of them will work.
3. function decorators will panic if any misused happened (eg req is not a struct type). I think there is no meaning to return a error because developer must change their code instead of handling there errors.

TODO:
- [x] document
- [x] better error message
- [x] bind request body
- [x] validate
- [x] form
- [x] multipart
- [x] Add some of old methods like GetHeaders etc. 
- [ ] Pointer support for struct fields.
- [ ] Add more tests and benchmarks.
- [ ] Better nested binding behavior like current binding
    - [x] Support `nested.and.age`, nested.age queries
    	- [x] Compiler support.
    	- [x] Decoder support.
    - [ ] Support `data.0.name` `data.0.users.0.name` and queries
    	- [x] Compiler support.
    	- [ ] Decoder support. 
    - [ ] Support `data[0][name]`, `nested[and][age]` queries (we can convert dots to square brackets or the opposite)

TODO for Extra Enhancements:
- [ ] Multipart file binding.
- [ ] Support map binding
- [ ] Default support.

### CODE REVIEW

**Comment by Trim21** [comment]
Do we need to support multiples content type automatically binding like the old way?  I'm thinking it's maybe ambiguous that handler may unmarshal some struct with with unexpected content type.

Like a struct with only XML or no tag at all may unmarshal json in this case, and some json like may map field case and some won't 🤔

**Comment by Trim21** [comment]
shoule we hide the original unmarshal message from `encoding.TextUnmarshaler.UnmarshalText` and `strconv.ParseInt`?

**Comment by efectn** [comment]
I reviewed this a bit today and this way seems good to me but i have a little question. Why don't we support map binding? I think we can add map support when checking the type in `compiler.go`

**Comment by efectn** [comment]
And we don't need default validator i think. Playground validator has too much and big sized indirect dependencies. But we can support custom validators by an struct validator interface.

**Comment by Trim21** [comment]
I wrap a `interface {Struct(s any) error}` in production code, it's only used in `example_test.go`, and go mod won't fetch test packages by default.

**Comment by Trim21** [comment]
> Why don't we support map binding

I think we can add old methods like `ctx.GetRespHeaders` back for this use case.

**Comment by Trim21** [comment]
The multipart and form body need a new decoder, I havn't done yet.

File: internal/reflectunsafe/reflectunsafe.go [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package reflectunsafe
+
+import (
+	"reflect"
+	"unsafe"
+)
+
+func ValueAndTypeID(v any) (reflect.Value, uintptr) {
```
Comment (efectn) [inline internal/reflectunsafe/reflectunsafe.go:8]: Maybe we can make this util?

File: binder.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+
+	if err := b.ctx.BodyXML(v); err != nil {
+		return b.setErr(err)
+	}
+
+	b.val = v
+	return b
+}
+
+func (b *Bind) Req(v any) *Bind {
```
Comment (efectn) [inline binder.go:77]: Maybe we can make this All(). And add other methods like Header, Cookie, Params etc.

File: binder_compile.go [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+var textUnmarshalerType = reflect.TypeOf((*encoding.TextUnmarshaler)(nil)).Elem()
+var bindUnmarshalerType = reflect.TypeOf((*Binder)(nil)).Elem()
+
+func compileReqParser(rt reflect.Type) (Decoder, error) {
+	var decoders []decoder
+
+	el := rt.Elem()
+	if el.Kind() != reflect.Struct {
+		panic("wrapped request need to struct")
```
Comment (efectn) [inline binder_compile.go:31]: I'll take a look here for map binding and multipart support.

File: binder_test.go [hunk 4]
```diff
@@ -0,0 +1,32 @@
+package fiber
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/require"
+	"github.com/valyala/fasthttp"
+)
+
+func Test_Binder(t *testing.T) {
```
Comment (efectn) [inline binder_test.go:10]: We need more tests and benchmarks

Comment (efectn) [reply binder_test.go:10]: You can add them like old ones.

File: ctx.go [hunk 5]
```diff
@@ -246,6 +264,14 @@ func (c *DefaultCtx) Body() []byte {
 	return body
 }
 
+func (c *DefaultCtx) BodyJSON(v any) error {
```
Comment (efectn) [inline ctx.go:267]: I think these methods can be a part of Bind() by chaining

**Review by efectn — COMMENTED** [review]
I added some comments. Maybe i can add some comments more in next days. Looks like great PR 🚀🚀

**Review by efectn — COMMENTED** [review]
Btw we can try to implement form, multipart form in this pr

File: binder_compile.go (hunk 3 above)
Comment (Trim21) [reply binder_compile.go:31]: Oh, this panic need to be "return error", I forget this

File: ctx.go (hunk 5 above)
Comment (Trim21) [reply ctx.go:267]: Binds already have json and XML method with extra content-type header checking

File: binder_test.go (hunk 4 above)
Comment (Trim21) [reply binder_test.go:10]: more tests are in bind_test.go

**Comment by Trim21** [comment]
> Btw we can try to implement form, multipart form in this pr

I add multipart, multipart files are not handled.

File: internal/reflectunsafe/reflectunsafe.go (hunk 1 above)
Comment (efectn) [reply internal/reflectunsafe/reflectunsafe.go:8]: I dont like putting much dirs in internal. Maybe we can put it inside of internal/bind. Wdym?

Comment (Trim21) [reply internal/reflectunsafe/reflectunsafe.go:8]: that't fine.

File: ctx.go (hunk 5 above)
Comment (efectn) [reply ctx.go:263]: I think you can add them into the Bind(). And you can check content type by Should/Must

Comment (Trim21) [reply ctx.go:263]: should we rename `Should/Must` to `Strict`?

Comment (efectn) [reply ctx.go]: Sounds good

**Comment by ReneWerner87** [comment]
@Trim21 can you update this again, will look at it next week, unfortunately had to do with the other issues and pull requests until now

thank you in any case already for the invested time

**Comment by efectn** [comment]
Hey @Trim21 can you check the discord

**Comment by ReneWerner87** [comment]
thank you for your work

this pull request is more complex than expected and i will probably need even longer for the review
i have already made some comments which you can please have a look at

my first impression is that we need a very good documentation here and please give me some examples that will help me to fully understand the possibilities of this customization

due to the fact that there are many possibilities here, we need to manage the consumer well so that there are few queries as there are many types and levels of developers

**Comment by ReneWerner87** [comment]
and in any case a lot of unittests to make sure with the mass of code lines that there are no or few errors in the logic and that we don't have to make any bugfix releases.

**Review by efectn — COMMENTED** [review]
We should improve nested binding. Maybe i can check it later.

**Comment by trim21** [comment]
> nested binding

can you explain this? like binding `?user.first_name=john&user.last_name=doe` to `struct { FirstName string; LastName string; } `?

**Comment by trim21** [comment]
> i have already made some comments which you can please have a look at

@ReneWerner87 do you forget to finally submit review comments in github ui? I can't see any review comments except your directly reply to the pull request

**Comment by ReneWerner87** [comment]
Yes sorry

File: binder.go [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+	if err := b.formDecode(v); err != nil {
+		return b.setErr(err)
+	}
+
+	b.val = v
+	return b
+}
+
+// Multipart unmarshal body as multipart/form-data
+// TODO: handle multipart files.
```
[the remark this answers is not in the document]
Comment (trim21) [reply binder.go:135]: No, we don't handle this in the original code, Have no idea how to bind multipart files, should we bind them to `io.Reader` or something?

File: bind_readme.md [hunk 7]
```diff
@@ -0,0 +1,191 @@
+# Fiber Binders
+
+Bind is new request/response binding feature for Fiber.
+By against old Fiber parsers, it supports custom binder registration,
+struct validation with high performance and easy to use.
```
[the remark this answers is not in the document]
Comment (trim21) [reply bind_readme.md:5]: there is a benchmark for 5 fields struct binding

```text
goos: windows
goarch: amd64
pkg: github.com/gofiber/fiber/v3
cpu: AMD Ryzen 7 PRO 4750U with Radeon Graphics
Benchmark_Bind_by_hand-16                               14412769                86.21 ns/op            0 B/op          0 allocs/op
Benchmark_Bind-16                                        4347765               275.0 ns/op            64 B/op          1 allocs/op
```

**Comment by trim21** [comment]
> and in any case a lot of unittests to make sure with the mass of code lines that there are no or few errors in the logic and that we don't have to make any bugfix releases.

I saw my PR again and looks like current all supported types are coveraged by tests already

File: bind_readme.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (ReneWerner87) [reply bind_readme.md:5]: well the commentary rather says that we have improved the old and new method of binding

not necessarily that here new and binding by hand is compared or ?

File: binder.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (ReneWerner87) [reply binder.go:135]: @efectn what do you think?

had we solved this somewhere before ? in the old method in v2 or in your implementation ?
or any other ideas which can help us here ?

File: bind_readme.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (trim21) [reply bind_readme.md:5]: > well the commentary rather says that we have improved the old and new method of binding
> 
> not necessarily that here new and binding by hand is compared or ?

here "high performance" is compared with old v2 binding (building map and use gorilla/schema)

https://github.com/gofiber/fiber/issues/2002#issuecomment-1208170445

File: binder.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (efectn) [reply binder.go:135]: We should check https://github.com/gin-gonic/gin/blob/master/binding/multipart_form_mapping.go#L35

**Comment by efectn** [comment]
I preapred a list of tasks. What do you think @trim21 
- Map binding (will check it asap)
- Multipart file binding 
- Better nested binding behavior like current binding
- Specific binding methods like current binding of v3
- More tests and benchmarks (i can do it)

**Comment by trim21** [comment]
I'm woking on multipart file binding now.

Why do we add a map binding insteading of adding a util function FasthttpArgsToMap?

**Comment by ReneWerner87** [comment]
> I'm woking on multipart file binding now.
> 
> Why do we add a map binding insteading of adding a util function FasthttpArgsToMap(interface{VisitAll})map[string]...?

sure good idea

or maybe we use just https://github.com/savsgio/dictpool 

can you test both ?

**Comment by trim21** [comment]
> sure good idea
> 
> or maybe we use just https://github.com/savsgio/dictpool
> 
> can you test both ?

I don't think github.com/savsgio/dictpool fit this case ...

**Comment by trim21** [comment]
> * Specific binding methods like current binding of v3

currently binding is controlled by struct tag, there is no need to add old API like Query, Params.

**Comment by efectn** [comment]
> I'm woking on multipart file binding now.
> 
> Why do we add a map binding insteading of adding a util function FasthttpArgsToMap(interface{VisitAll})map[string]...?

I think it would be more useful to adding map binding instead of an util. Otherwise we have to add some methods to shortify getting them.

> > * Specific binding methods like current binding of v3
> 
> currently binding is controlled by struct tag, there is no need to add old API like Query, Params.

Oh, right. Controlling by struct tag seems ok

File: bind_readme.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (ReneWerner87) [reply bind_readme.md:5]: thx this is a really good info

**Comment by ReneWerner87** [comment]
thanks for the patience with my review and the work you put in here ❤️

please update the branch again and just let me know when the pull request is ready for further review

**Comment by trim21** [comment]
Got little busy these days, will update it soon

**Comment by efectn** [comment]
I think we should add GetHeaders, GetRespHeaders, GetParams back. They are useful for something needs map. Should we add them into the ctx or Bind (like Bind().GetHeaders())? @ReneWerner87 @trim21

**Comment by efectn** [comment]
Hi @trim21. I added simple nested binding support. Can you review my last commit?

I also plan to add support for slices as soon as possible.

**Comment by FGYFFFF** [comment]
Hello~ The idea of implementing v3's bind is really great! I have an idea I'd like to discuss.
```
	type Req struct {
		F1  *string `param:"f1" query:"f1" header:"f1"`
	}
```
the binding order is: "query"、 "param" 、"header".

If the 'query' missed the bind, the field will not be decoded.
So is it possible to continue trying to bind the "param" parameter after not binding to the query parameter. This is a "best effort delivery" concept.

In addition, I think this 'pr' is very meaningful, if necessary I can also participate in

**Comment by efectn** [comment]
> Hello~ The idea of implementing v3's bind is really great! I have an idea I'd like to discuss.
> 
> ```
> 	type Req struct {
> 		F1  *string `param:"f1" query:"f1" header:"f1"`
> 	}
> ```
> 
> the binding order is: "query"、 "param" 、"header".
> 
> If the 'query' missed the bind, the field will not be decoded. So is it possible to continue trying to bind the "param" parameter after not binding to the query parameter. This is a "best effort delivery" concept.
> 
> In addition, I think this 'pr' is very meaningful, if necessary I can also participate in

Hi @FGYFFFF. You can help to complete this PR. I don't have much time nowadays so i couldn't find time to finish the PR. Feel free to help us <3

**Comment by FGYFFFF** [comment]
@trim21 Can you give me access to a push code? I want to add support for the "float" type.

**Comment by FGYFFFF** [comment]
hey, all ~ @trim21 @efectn , If the field of the structure is an array, do we need to support it?

**Comment by trim21** [comment]
> hey, all ~ @trim21 @efectn , If the field of the structure is an array, do we need to support it?

it's already supported in this PR I think...

https://github.com/gofiber/fiber/blob/081809e8a69649fafb59fa11dd629e9d00117a7a/bind_readme.md#quick-start

**Comment by FGYFFFF** [comment]
```
	var req struct {
		Data []struct {
			Name string `query:"name"`
			Age  int    `query:"age"`
		} `query:"data"`
	}
```
I read the source code, "Slice" already supports it, but "Array" doesn't seem to support it yet. like blow
```
	type Req struct {
		ID1 [5]string `param:"id"`
	}
```
will occur a error for “unsupported type [5]string”.

**Comment by trim21** [comment]
> ```
> 	var req struct {
> 		Data []struct {
> 			Name string `query:"name"`
> 			Age  int    `query:"age"`
> 		} `query:"data"`
> 	}
> ```
> 
> I read the source code, "Slice" already supports it, but "Array" doesn't seem to support it yet. like blow
> 
> ```
> 	type Req struct {
> 		ID1 [5]string `query:"id"`
> 	}
> ```
> 
> will occur a error for “unsupported type [5]string”. @trim21

Oh, sorry... I'm think slice .....

**Comment by trim21** [comment]
> ```
> 	var req struct {
> 		Data []struct {
> 			Name string `query:"name"`
> 			Age  int    `query:"age"`
> 		} `query:"data"`
> 	}
> ```
> 
> I read the source code, "Slice" already supports it, but "Array" doesn't seem to support it yet. like blow
> 
> ```
> 	type Req struct {
> 		ID1 [5]string `query:"id"`
> 	}
> ```
> 
> will occur a error for “unsupported type [5]string”. @trim21

I don't think it's a bad idea, but should it return a validation error when trying to bind `q=1&q=2` to `[5]int`?

**Comment by FGYFFFF** [comment]
> > ```
> > 	var req struct {
> > 		Data []struct {
> > 			Name string `query:"name"`
> > 			Age  int    `query:"age"`
> > 		} `query:"data"`
> > 	}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I read the source code, "Slice" already supports it, but "Array" doesn't seem to support it yet. like blow
> > ```
> > 	type Req struct {
> > 		ID1 [5]string `query:"id"`
> > 	}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > will occur a error for “unsupported type [5]string”. @trim21
> 
> I don't think it's a bad idea, but should it return a validation error when trying to bind `q=1&q=2` to `[5]int`?

yes~ like gin (https://github.com/gin-gonic/gin/blob/3010cbd7f4eccdbb610c510274895e083b8c058c/binding/form_mapping.go#L184), it will determine the length of the array in advance.  
If possible, I can finish this FEATURE

**Comment by trim21** [comment]
> yes~ like gin (https://github.com/gin- gonic/gin/blob/3010cbd7f4eccdbb610c510274895e083b8c058c/binding/form_mapping.go#L184), it will determine the length of the array in advance.
> If possible, I can finish this FEATURE

I'm not against this feature or you adding it in this PR, but if possible I'd like to get this PR merged first...

**Comment by FGYFFFF** [comment]
> 

I'm not trying to grab this PR, I just found some problems in the process of research, so I raised it.
Sorry, maybe because my description was not clear

**Comment by FGYFFFF** [comment]
> to get this PR merged first

I'm not trying to grab this PR, I just found some problems in the process of research, so I raised it.
Sorry, maybe because my description was not clear

**Comment by trim21** [comment]
> > to get this PR merged first
> 
> I'm not trying to grab this PR, I just found some problems in the process of research, so I raised it. Sorry, maybe because my description was not clear

No... I didn't mean that. That's ok if you could help to add feature or finish this PR, or even just re-open a new PR based on this one, it's not a problem to me

**Comment by FGYFFFF** [comment]
> 

ok~😄😄😄

**Comment by FGYFFFF** [comment]
So far, I have found some feature that need to be added in:
* Array
* multiple pointer (gin supported, does fiber need to support?)

I will complete these FEATURES in my free time.

**Comment by ReneWerner87** [comment]
@FGYFFFF how far are you ?

**Comment by ReneWerner87** [comment]
ok perfect, thx

**Review by skydiver1980 — APPROVED** [review]

**Comment by ReneWerner87** [comment]
@trim21  @efectn 
what tasks are still open here? we can now pay more attention to v3 

can you merge again ? and document the last open points so that we can actively help here or find decisions

**Comment by ReneWerner87** [comment]
todos: 

- [ ] Nested structs are supported. However nested slice structs are not supported. slice field decoder needs to be rewrited for that case
- [ ] We need to make more check for pointer usage cases
- [ ] we need to support files

**Comment by ReneWerner87** [comment]
@coderabbitai can you suggest solutions for the open todos

Can you also check the existing code

**Pull request closed without merging**